File: apps/api/src/common/utils/response.ts
/**
 * 统一响应格式封装
 * 所有 API 响应统一 { code, message, data } 格式
 */
import { FastifyReply } from 'fastify';
import { ERROR_CODES } from '@alldata/shared/constants/index.js';

export interface SuccessResponse<T = unknown> {
  code: number;
  message: string;
  data: T;
}

export interface ErrorResponse {
  code: number;
  message: string;
  data: null;
}

/** 成功响应 */
export function sendSuccess<T>(reply: FastifyReply, data: T, message = 'success', code = 200): void {
  reply.code(code).send({ code, message, data });
}

/** 分页响应 */
export function sendPaginated<T>(
  reply: FastifyReply,
  list: T[],
  total: number,
  page: number,
  pageSize: number,
): void {
  const totalPage = Math.ceil(total / pageSize) || 1;
  reply.send({
    code: 200,
    message: 'success',
    data: {
      list,
      page_info: {
        current_page: page,
        page_size: pageSize,
        total_page: totalPage,
        total,
      },
    },
  });
}

/** 错误响应 */
export function sendError(
  reply: FastifyReply,
  code: number,
  message: string,
  httpStatus?: number,
): void {
  reply.code(httpStatus ?? code).send({ code, message, data: null });
}

/** 常用错误快捷方法 */
export const ApiError = {
  badRequest: (reply: FastifyReply, msg = '参数错误') =>
    sendError(reply, ERROR_CODES.BAD_REQUEST, msg, 400),

  unauthorized: (reply: FastifyReply, msg = '未授权，请登录') =>
    sendError(reply, ERROR_CODES.UNAUTHORIZED, msg, 401),

  forbidden: (reply: FastifyReply, msg = '权限不足') =>
    sendError(reply, ERROR_CODES.FORBIDDEN, msg, 403),

  notFound: (reply: FastifyReply, msg = '资源不存在') =>
    sendError(reply, ERROR_CODES.NOT_FOUND, msg, 404),

  tooMany: (reply: FastifyReply, msg = '请求频率超限') =>
    sendError(reply, ERROR_CODES.TOO_MANY_REQUESTS, msg, 429),

  tokenExpired: (reply: FastifyReply) =>
    sendError(reply, ERROR_CODES.TOKEN_EXPIRED, 'Token 已过期', 401),

  dataPermission: (reply: FastifyReply) =>
    sendError(reply, ERROR_CODES.DATA_PERMISSION_DENIED, '数据权限不足', 403),

  internal: (reply: FastifyReply, msg = '服务器内部错误') =>
    sendError(reply, ERROR_CODES.INTERNAL_ERROR, msg, 500),
};

File: apps/api-go/pkg/response/response.go
package response

import (
	"net/http"

	"github.com/gin-gonic/gin"
)

// Response 统一响应格式
type Response struct {
	Code    int         `json:"code"`
	Message string      `json:"message"`
	Data    interface{} `json:"data"`
}

// Success 成功响应
func Success(c *gin.Context, data interface{}) {
	c.JSON(http.StatusOK, Response{
		Code:    200,
		Message: "success",
		Data:    data,
	})
}

// Created 创建成功
func Created(c *gin.Context, data interface{}) {
	c.JSON(http.StatusCreated, Response{
		Code:    201,
		Message: "created",
		Data:    data,
	})
}

// Error 错误响应
func Error(c *gin.Context, code int, message string) {
	c.JSON(code, Response{
		Code:    code,
		Message: message,
		Data:    nil,
	})
}

// BadRequest 400 错误
func BadRequest(c *gin.Context, message string) {
	Error(c, 400, message)
}

// Unauthorized 401 错误
func Unauthorized(c *gin.Context, message string) {
	Error(c, 401, message)
}

// Forbidden 403 错误
func Forbidden(c *gin.Context, message string) {
	Error(c, 403, message)
}

// NotFound 404 错误
func NotFound(c *gin.Context, message string) {
	Error(c, 404, message)
}

// InternalError 500 错误
func InternalError(c *gin.Context, message string) {
	Error(c, 500, message)
}

// Paginated 分页响应
type PaginatedData struct {
	List     interface{} `json:"list"`
	PageInfo PageInfo    `json:"page_info"`
}

type PageInfo struct {
	CurrentPage int   `json:"current_page"`
	PageSize    int   `json:"page_size"`
	TotalPage   int   `json:"total_page"`
	Total       int64 `json:"total"`
}

func Paginate(c *gin.Context, list interface{}, total int64, page, pageSize int) {
	totalPage := int(total) / pageSize
	if int(total)%pageSize > 0 {
		totalPage++
	}
	c.JSON(http.StatusOK, Response{
		Code:    200,
		Message: "success",
		Data: PaginatedData{
			List: list,
			PageInfo: PageInfo{
				CurrentPage: page,
				PageSize:    pageSize,
				TotalPage:   totalPage,
				Total:       total,
			},
		},
	})
}
